File: UnionFindSet/src/main/python/solution547.py
from typing import List


class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:

        n = len(isConnected)

        self.fa = [i for i in range(n)]

        for i in range(n):
            for j in range(n):
                if isConnected[i][j]:
                    self.union_set(i, j)

        ans = 0

        for i in range(n):
            if self.find(i) == i:
                ans += 1

        return ans

    def find(self, x):
        if x == self.fa[x]:
            return x

        self.fa[x] = self.find(self.fa[x])

        return self.fa[x]

    def union_set(self, x, y):
        x = self.find(x)
        y = self.find(y)

        if x != y:
            self.fa[x] = y


s = Solution()
isConnected = [[1, 1, 0], [1, 1, 0], [0, 0, 1]]
res = s.findCircleNum(isConnected)
print(res)

fa = [0, 1, 1, 2, 3, 4]


def test_find(x):

    if x == fa[x]:
        return x

    fa[x] = test_find(fa[x])

    print(fa)
    return fa[x]


print(test_find(5))

File: UnionFindSet/src/main/java/Solution547.java
public class Solution547 {
    int[] fa;

    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        fa = new int[n];

        for (int i = 0; i < n; i++) {
            fa[i] = i;
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (isConnected[i][j] == 1) {
                    unionSet(i, j);
                }
            }
        }

        int ans = 0;

        for (int i = 0; i < n; i++) {
            if (fa[i] == i) {
                ans++;
            }
        }

        return ans;

    }

    private int find(int x) {
        if (x == fa[x]) {
            return x;
        }

        return fa[x] = find(fa[x]);
    }

    private void unionSet(int x, int y) {
        x = find(x);
        y = find(y);

        if (x != y) {
            fa[x] = y;
        }
    }
}
